### PR TITLE
Correct client certificate validation policy on Windows and macOS.

### LINKED ISSUE #1803: Client Certificate Support for OpenSSL
#1243 added basic support for client certificates on Windows with Schannel.
We should get parity on Linux/OpenSSL as well.

### PR DESCRIPTION
**Pull request by anrossi**
The wrong certificate validation policy was being used for client certificates with OpenSSL on Windows and macOS. Closes #1803.

### CODE REVIEW

File: src/platform/cert_capi.c [hunk 1]
```diff
@@ -801,9 +801,10 @@ CxPlatCertVerifyCertChainPolicy(
 
     memset(&HttpsPolicy, 0, sizeof(HTTPSPolicyCallbackData));
     HttpsPolicy.cbStruct = sizeof(HTTPSPolicyCallbackData);
-    HttpsPolicy.dwAuthType = AUTHTYPE_SERVER;
+    HttpsPolicy.dwAuthType =
+        (CredFlags & QUIC_CREDENTIAL_FLAG_CLIENT) ? AUTHTYPE_SERVER : AUTHTYPE_CLIENT;
     HttpsPolicy.fdwChecks = 0;
-    HttpsPolicy.pwszServerName = ServerName;
+    HttpsPolicy.pwszServerName = (CredFlags & QUIC_CREDENTIAL_FLAG_CLIENT) ? NULL : ServerName;
```
Comment (nibanks) [inline src/platform/cert_capi.c:807]: This is reversed, isn't it? Clients should provide the server name.

File: src/platform/cert_capi.c [hunk 2]
```diff
@@ -873,21 +874,27 @@ CxPlatCertValidateChain(
...
 
     memset(&ChainPara, 0, sizeof(ChainPara));
     ChainPara.cbSize = sizeof(ChainPara);
     ChainPara.RequestedUsage.dwType = USAGE_MATCH_TYPE_OR;
-    ChainPara.RequestedUsage.Usage.cUsageIdentifier = ARRAYSIZE(UsageOids);
-    ChainPara.RequestedUsage.Usage.rgpszUsageIdentifier = (LPSTR*)UsageOids;
+    ChainPara.RequestedUsage.Usage.cUsageIdentifier =
+        (CredFlags & QUIC_CREDENTIAL_FLAG_CLIENT) ? ARRAYSIZE(ClientUsageOids) : ARRAYSIZE(ServerUsageOids);
+    ChainPara.RequestedUsage.Usage.rgpszUsageIdentifier =
+        (CredFlags & QUIC_CREDENTIAL_FLAG_CLIENT) ? (LPSTR*)ClientUsageOids : (LPSTR*)ServerUsageOids;
```
Comment (nibanks) [inline src/platform/cert_capi.c:897]: Are these swapped too? Clients should validate the certs are used for server auth, right?

File: src/platform/darwin_openssl.c [hunk 3]
```diff
@@ -123,7 +123,9 @@ CxPlatTlsVerifyCertificate(
         goto Exit;
     }
 
-    SSLPolicy = SecPolicyCreateSSL(true, SNIString);
+    SSLPolicy = SecPolicyCreateSSL(
+        (CredFlags & QUIC_CREDENTIAL_FLAG_CLIENT) ? false : true,
+        SNIString);
```
Comment (nibanks) [inline src/platform/darwin_openssl.c:128]: nit: when the arguments are on a different line than the function call, please put the function call on its own line too.
```suggestion
    SSLPolicy =
        SecPolicyCreateSSL(
            (CredFlags & QUIC_CREDENTIAL_FLAG_CLIENT) ? FALSE : TRUE,
            SNIString);
```
Also, this is `c` code, so `TRUE` or `FALSE` please.

File: src/platform/cert_capi.c [hunk 4]
```diff
@@ -873,21 +874,27 @@ CxPlatCertValidateChain(
...
         *ValidationError = NO_ERROR;
     }
 
     memset(&ChainPara, 0, sizeof(ChainPara));
     ChainPara.cbSize = sizeof(ChainPara);
     ChainPara.RequestedUsage.dwType = USAGE_MATCH_TYPE_OR;
-    ChainPara.RequestedUsage.Usage.cUsageIdentifier = ARRAYSIZE(UsageOids);
-    ChainPara.RequestedUsage.Usage.rgpszUsageIdentifier = (LPSTR*)UsageOids;
+    ChainPara.RequestedUsage.Usage.cUsageIdentifier =
+        (CredFlags & QUIC_CREDENTIAL_FLAG_CLIENT) ?  ARRAYSIZE(ServerUsageOids) : ARRAYSIZE(ClientUsageOids);
```
Comment (nibanks) [inline src/platform/cert_capi.c:895]:
```suggestion
        (CredFlags & QUIC_CREDENTIAL_FLAG_CLIENT) ? ARRAYSIZE(ServerUsageOids) : ARRAYSIZE(ClientUsageOids);
```

File: src/platform/darwin_openssl.c [hunk 5]
```diff
@@ -123,7 +123,10 @@ CxPlatTlsVerifyCertificate(
         goto Exit;
     }
 
-    SSLPolicy = SecPolicyCreateSSL(true, SNIString);
+    SSLPolicy =
+        SecPolicyCreateSSL(
+            (CredFlags & QUIC_CREDENTIAL_FLAG_CLIENT) ? FALSE : TRUE,
```
Comment (thhous-msft) [inline src/platform/darwin_openssl.c:128]: This needs to be reversed too. For client side (verifying a server cert) this needs to be true. Also, this is where the merge conflict with main in occuring so it won't run until main is merged back in.

**Comment by ThadHouse** [comment]
This change passing this easily has me worried. Its making me wonder if somehow the right tests are not being ran in openssl.  Something we should definitely double check.

**Comment by nibanks** [comment]
> This change passing this easily has me worried. Its making me wonder if somehow the right tests are not being ran in openssl. Something we should definitely double check.

I also agree with this sentiment. We need to take a close look at the tests (with this PR) to (1) make sure the existing tests are working and (2) add any tests that we are missing to actually validate this. The fact that tests passed without this change is bad.

**Comment by nibanks** [comment]
Also, CLOG failed, so you need to regen the sidecar/headers.

File: src/platform/cert_capi_openssl.c [hunk 6]
```diff
@@ -87,15 +87,12 @@ CxPlatTlsVerifyCertificate(
         CertFlags |= CERT_CHAIN_REVOCATION_CHECK_CHAIN_EXCLUDE_ROOT;
     }
 
-    uint32_t IgnoreFlags =
```
Comment (anrossi) [inline src/platform/cert_capi_openssl.c:90]: I feel like I did this on purpose when I wrote it, but I don't remember now, so I'm removing it.

**Review by nibanks — APPROVED** [review]

**Review by thhous-msft — APPROVED** [review]

**Pull request merged**